File: connectors/javascript/aurakeeper.js
(function (globalScope, factory) {
  var exported = factory();

  if (typeof module === "object" && module.exports) {
    module.exports = exported;
    return;
  }

  globalScope.AuraKeeper = exported;
})(typeof globalThis !== "undefined" ? globalThis : this, function () {
  "use strict";

  function AuraKeeperConnector(options) {
    if (!options || !options.endpoint) {
      throw new Error("AuraKeeperConnector requires an endpoint.");
    }

    if (!options.apiToken) {
      throw new Error("AuraKeeperConnector requires an apiToken.");
    }

    if (!options.serviceName) {
      throw new Error("AuraKeeperConnector requires a serviceName.");
    }

    this.options = options;
    this.pendingRequests = new Set();
    this.installed = false;
    this.listeners = {};
  }

  AuraKeeperConnector.prototype.install = function install() {
    if (this.installed) {
      return this;
    }

    if (isBrowserRuntime() && this.options.captureBrowser !== false) {
      this.listeners.browserError = this.handleBrowserError.bind(this);
      this.listeners.browserRejection = this.handleBrowserRejection.bind(this);

      globalThis.addEventListener("error", this.listeners.browserError);
      globalThis.addEventListener(
        "unhandledrejection",
        this.listeners.browserRejection
      );
    }

    if (isNodeRuntime() && this.options.captureNode !== false) {
      this.listeners.nodeException = this.handleNodeException.bind(this);
      this.listeners.nodeRejection = this.handleNodeRejection.bind(this);

      process.on("uncaughtException", this.listeners.nodeException);
      process.on("unhandledRejection", this.listeners.nodeRejection);
    }

    this.installed = true;
    return this;
  };

  AuraKeeperConnector.prototype.uninstall = function uninstall() {
    if (!this.installed) {
      return this;
    }

    if (this.listeners.browserError) {
      globalThis.removeEventListener("error", this.listeners.browserError);
    }

    if (this.listeners.browserRejection) {
      globalThis.removeEventListener(
        "unhandledrejection",
        this.listeners.browserRejection
      );
    }

    if (this.listeners.nodeException) {
      removeProcessListener("uncaughtException", this.listeners.nodeException);
    }

    if (this.listeners.nodeRejection) {
      removeProcessListener("unhandledRejection", this.listeners.nodeRejection);
    }

    this.listeners = {};
    this.installed = false;
    return this;
  };

  AuraKeeperConnector.prototype.captureException = function captureException(
    error,
    overrides
  ) {
    var payload = this.buildPayload(error, overrides || {});

    if (!payload) {
      return Promise.resolve(null);
    }

    return this.track(this.send(payload));
  };

  AuraKeeperConnector.prototype.captureMessage = function captureMessage(
    message,
    overrides
  ) {
    var syntheticError = new Error(String(message));
    syntheticError.name = "Error";

    return this.captureException(syntheticError, overrides);
  };

  AuraKeeperConnector.prototype.flush = function flush() {
    return Promise.allSettled(Array.from(this.pendingRequests));
  };

  AuraKeeperConnector.prototype.handleBrowserError = function handleBrowserError(
    event
  ) {
    var error =
      event && event.error
        ? event.error
        : new Error((event && event.message) || "Unhandled browser error");
    var details = {
      filename: event && event.filename ? event.filename : undefined,
      line: event && typeof event.lineno === "number" ? event.lineno : undefined,
      column:
        event && typeof event.colno === "number" ? event.colno : undefined,
    };

    this.captureException(error, {
      handled: false,
      level: "error",
      platform: this.options.platform || "web",
      source: {
        runtime: "browser",
      },
      details: compactObject(details),
    }).catch(this.handleTransportFailure.bind(this));
  };

  AuraKeeperConnector.prototype.handleBrowserRejection =
    function handleBrowserRejection(event) {
      var normalized = normalizeUnknownError(
        event ? event.reason : undefined,
        "Unhandled promise rejection"
      );

      this.captureException(normalized.error, {
        handled: false,
        level: "error",
        platform: this.options.platform || "web",
        source: {
          runtime: "browser",
        },
        details: mergeObjects(normalized.details, {
          rejection: true,
        }),
      }).catch(this.handleTransportFailure.bind(this));
    };

  AuraKeeperConnector.prototype.handleNodeException = function handleNodeException(
    error
  ) {
    this.captureException(error, {
      handled: false,
      level: "critical",
      platform: this.options.platform || "backend",
      source: {
        runtime: "node",
      },
    }).catch(this.handleTransportFailure.bind(this));
  };

  AuraKeeperConnector.prototype.handleNodeRejection = function handleNodeRejection(
    reason
  ) {
    var normalized = normalizeUnknownError(reason, "Unhandled promise rejection");

    this.captureException(normalized.error, {
      handled: false,
      level: "error",
      platform: this.options.platform || "backend",
      source: {
        runtime: "node",
      },
      details: mergeObjects(normalized.details, {
        rejection: true,
      }),
    }).catch(this.handleTransportFailure.bind(this));
  };

  AuraKeeperConnector.prototype.handleTransportFailure =
    function handleTransportFailure(error) {
      if (typeof this.options.onTransportError === "function") {
        this.options.onTransportError(error);
        return;
      }

      if (typeof console !== "undefined" && console.error) {
        console.error("AuraKeeper failed to send error log.", error);
      }
    };

  AuraKeeperConnector.prototype.buildPayload = function buildPayload(
    error,
    overrides
  ) {
    var normalized = normalizeUnknownError(error, "Unknown error");
    var mergedDetails = mergeObjects(normalized.details, overrides.details);
    var mergedContext = this.buildContext(overrides);
    var payload = {
      eventId: overrides.eventId || generateEventId(),
      occurredAt: overrides.occurredAt || new Date().toISOString(),
      level: overrides.level || "error",
      platform: overrides.platform || this.options.platform || detectPlatform(),
      environment: overrides.environment || this.options.environment,
      service: compactObject(
        mergeObjects(
          {
            name: this.options.serviceName,
            version: this.options.serviceVersion,
            instanceId: this.options.instanceId,
          },
          overrides.service
        )
      ),
      source: compactObject(
        mergeObjects(
          {
            runtime: detectRuntime(),
            language: this.options.language || "javascript",
            framework: this.options.framework,
            component: this.options.component,
          },
          overrides.source
        )
      ),
      error: compactObject({
        type: overrides.type || normalized.error.name || "Error",
        message:
          overrides.message || normalized.error.message || "Unknown error",
        code: overrides.code || readErrorCode(error),
        stack: overrides.stack || normalized.error.stack,
        handled:
          typeof overrides.handled === "boolean" ? overrides.handled : true,
        details: hasOwnKeys(mergedDetails) ? mergedDetails : undefined,
      }),
      context: hasOwnKeys(mergedContext) ? mergedContext : undefined,
    };

    if (typeof this.options.beforeSend === "function") {
      var nextPayload = this.options.beforeSend(payload);

      if (nextPayload === false || nextPayload == null) {
        return null;
      }

      payload = nextPayload;
    }

    return pruneEmpty(compactObject(payload));
  };

  AuraKeeperConnector.prototype.buildContext = function buildContext(overrides) {
    var optionContext = this.options.context || {};
    var overrideContext = overrides.context || {};
    var tags = []
      .concat(this.options.tags || [])
      .concat(optionContext.tags || [])
      .concat(overrideContext.tags || [])
      .concat(overrides.tags || []);

    return pruneEmpty(
      mergeObjects(optionContext, overrideContext, {
        request: mergeObjects(
          optionContext.request,
          overrideContext.request,
          overrides.request
        ),
        user: mergeObjects(optionContext.user, overrideContext.user, overrides.user),
        session: mergeObjects(
          optionContext.session,
          overrideContext.session,
          overrides.session
        ),
        device: mergeObjects(
          optionContext.device,
          overrideContext.device,
          overrides.device
        ),
        correlationId:
          overrides.correlationId ||
          overrideContext.correlationId ||
          optionContext.correlationId,
        tags: tags.length ? uniqueStrings(tags) : undefined,
      })
    );
  };

  AuraKeeperConnector.prototype.send = function send(payload) {
    var transport = this.options.transport || defaultTransport;

    return Promise.resolve(
      transport({
        endpoint: this.options.endpoint,
        apiToken: this.options.apiToken,
        payload: payload,
        headers: this.options.headers || {},
        fetch: this.options.fetch,
      })
    );
  };

  AuraKeeperConnector.prototype.track = function track(promise) {
    var pending = Promise.resolve(promise).finally(
      function cleanup() {
        this.pendingRequests.delete(pending);
      }.bind(this)
    );

    this.pendingRequests.add(pending);
    return pending;
  };

  function createAuraKeeperConnector(options) {
    return new AuraKeeperConnector(options);
  }

  function defaultTransport(config) {
    var fetchImpl = config.fetch || globalThis.fetch;

    if (typeof fetchImpl !== "function") {
      throw new Error(
        "AuraKeeper requires fetch. Provide options.fetch or options.transport when fetch is unavailable."
      );
    }

    return fetchImpl(config.endpoint, {
      method: "POST",
      headers: mergeObjects(
        {
          "content-type": "application/json",
          "X-API-Token": config.apiToken,
        },
        config.headers
      ),
      body: JSON.stringify(config.payload),
      keepalive: true,
    }).then(function handleResponse(response) {
      if (response && response.ok) {
        return response;
      }

      if (!response || typeof response.text !== "function") {
        throw new Error("AuraKeeper request failed.");
      }

      return response.text().then(function onResponseText(text) {
        throw new Error(
          "AuraKeeper request failed with status " +
            response.status +
            ": " +
            text
        );
      });
    });
  }

  function normalizeUnknownError(value, fallbackMessage) {
    if (value instanceof Error) {
      return {
        error: value,
        details: undefined,
      };
    }

    if (isObject(value) && typeof value.message === "string") {
      var errorLike = new Error(value.message);
      errorLike.name =
        typeof value.name === "string" && value.name ? value.name : "Error";
      errorLike.stack =
        typeof value.stack === "string" ? value.stack : errorLike.stack;

      return {
        error: errorLike,
        details: sanitizeJson(value),
      };
    }

    var message =
      typeof value === "string" && value ? value : fallbackMessage || "Unknown error";
    var normalizedError = new Error(message);

    return {
      error: normalizedError,
      details:
        typeof value === "undefined"
          ? undefined
          : {
              originalValue: sanitizeJson(value),
            },
    };
  }

  function readErrorCode(error) {
    if (error && typeof error.code === "string") {
      return error.code;
    }

    return undefined;
  }

  function detectPlatform() {
    if (isBrowserRuntime()) {
      return "web";
    }

    if (isNodeRuntime()) {
      return "backend";
    }

    return "worker";
  }

  function detectRuntime() {
    if (isBrowserRuntime()) {
      return "browser";
    }

    if (isNodeRuntime()) {
      return "node";
    }

    return "javascript";
  }

  function isBrowserRuntime() {
    return (
      typeof globalThis !== "undefined" &&
      typeof globalThis.addEventListener === "function" &&
      typeof globalThis.document !== "undefined"
    );
  }

  function isNodeRuntime() {
    return (
      typeof process !== "undefined" &&
      process &&
      typeof process.on === "function" &&
      process.release &&
      process.release.name === "node"
    );
  }

  function generateEventId() {
    if (
      typeof globalThis !== "undefined" &&
      globalThis.crypto &&
      typeof globalThis.crypto.randomUUID === "function"
    ) {
      return globalThis.crypto.randomUUID();
    }

    return "evt_" + Math.random().toString(16).slice(2) + Date.now().toString(16);
  }

  function removeProcessListener(eventName, listener) {
    if (typeof process.off === "function") {
      process.off(eventName, listener);
      return;
    }

    if (typeof process.removeListener === "function") {
      process.removeListener(eventName, listener);
    }
  }

  function uniqueStrings(values) {
    var output = [];
    var seen = new Set();

    values.forEach(function each(value) {
      if (typeof value !== "string" || !value) {
        return;
      }

      if (seen.has(value)) {
        return;
      }

      seen.add(value);
      output.push(value);
    });

    return output;
  }

  function compactObject(value) {
    if (!isObject(value)) {
      return value;
    }

    var output = {};

    Object.keys(value).forEach(function eachKey(key) {
      if (typeof value[key] !== "undefined") {
        output[key] = value[key];
      }
    });

    return output;
  }

  function pruneEmpty(value) {
    if (Array.isArray(value)) {
      var nextArray = value
        .map(function eachItem(entry) {
          return pruneEmpty(entry);
        })
        .filter(function filterEmpty(entry) {
          return typeof entry !== "undefined";
        });

      return nextArray.length ? nextArray : undefined;
    }

    if (!isObject(value)) {
      return value;
    }

    var nextObject = {};

    Object.keys(value).forEach(function eachKey(key) {
      var nextValue = pruneEmpty(value[key]);

      if (typeof nextValue !== "undefined") {
        nextObject[key] = nextValue;
      }
    });

    return Object.keys(nextObject).length ? nextObject : undefined;
  }

  function hasOwnKeys(value) {
    return isObject(value) && Object.keys(value).length > 0;
  }

  function mergeObjects() {
    var merged = {};

    for (var index = 0; index < arguments.length; index += 1) {
      var current = arguments[index];

      if (!isObject(current)) {
        continue;
      }

      Object.keys(current).forEach(function eachKey(key) {
        if (typeof current[key] !== "undefined") {
          merged[key] = current[key];
        }
      });
    }

    return merged;
  }

  function isObject(value) {
    return value != null && typeof value === "object" && !Array.isArray(value);
  }

  function sanitizeJson(value) {
    return sanitizeJsonValue(value, new Set(), 0);
  }

  function sanitizeJsonValue(value, seen, depth) {
    if (value == null) {
      return value;
    }

    if (depth >= 6) {
      return "[MaxDepth]";
    }

    if (
      typeof value === "string" ||
      typeof value === "number" ||
      typeof value === "boolean"
    ) {
      return value;
    }

    if (typeof value === "bigint") {
      return value.toString();
    }

    if (typeof value === "function") {
      return "[Function]";
    }

    if (value instanceof Error) {
      return compactObject({
        name: value.name,
        message: value.message,
        stack: value.stack,
        code: readErrorCode(value),
      });
    }

    if (Array.isArray(value)) {
      return value.map(function eachArrayItem(entry) {
        return sanitizeJsonValue(entry, seen, depth + 1);
      });
    }

    if (!isObject(value)) {
      return String(value);
    }

    if (seen.has(value)) {
      return "[Circular]";
    }

    seen.add(value);

    var output = {};

    Object.keys(value).forEach(function eachObjectKey(key) {
      output[key] = sanitizeJsonValue(value[key], seen, depth + 1);
    });

    seen.delete(value);
    return output;
  }

  return {
    AuraKeeperConnector: AuraKeeperConnector,
    createAuraKeeperConnector: createAuraKeeperConnector,
  };
});

File: connectors/ruby/lib/aurakeeper.rb
require "date"
require "json"
require "net/http"
require "securerandom"
require "time"
require "uri"

module AuraKeeper
  VERSION = "0.1.0"

  def self.create_aurakeeper_connector(**kwargs)
    Connector.new(**kwargs)
  end

  class Connector
    MAX_SANITIZE_DEPTH = 6

    def initialize(
      endpoint:,
      api_token:,
      service_name:,
      service_version: nil,
      environment: nil,
      platform_name: nil,
      framework: nil,
      component: nil,
      instance_id: nil,
      tags: nil,
      context: nil,
      headers: nil,
      timeout: 5.0,
      transport: nil,
      before_send: nil,
      on_transport_error: nil,
      capture_uncaught: true
    )
      raise ArgumentError, "AuraKeeper::Connector requires an endpoint." if blank?(endpoint)
      raise ArgumentError, "AuraKeeper::Connector requires an api_token." if blank?(api_token)
      raise ArgumentError, "AuraKeeper::Connector requires a service_name." if blank?(service_name)

      @endpoint = endpoint
      @api_token = api_token
      @service_name = service_name
      @service_version = service_version
      @environment = environment
      @platform_name = platform_name
      @framework = framework
      @component = component
      @instance_id = instance_id
      @tags = Array(tags).compact
      @context = context.is_a?(Hash) ? context.dup : {}
      @headers = headers.is_a?(Hash) ? headers.dup : {}
      @timeout = timeout
      @transport = transport || method(:default_transport)
      @before_send = before_send
      @on_transport_error = on_transport_error
      @capture_uncaught = capture_uncaught
      @installed = false
      @uncaught_hook_registered = false
    end

    def install
      register_uncaught_hook if @capture_uncaught
      @installed = true
      self
    end

    def uninstall
      @installed = false
      self
    end

    def close
      uninstall
    end

    def flush
      []
    end

    def capture_exception(error, overrides = nil, **kwargs)
      payload = build_payload(error, overrides, **kwargs)
      return nil if payload.nil?

      deliver(payload)
    end

    def capture_message(message, overrides = nil, **kwargs)
      capture_exception(RuntimeError.new(String(message)), overrides, **kwargs)
    end

    def build_payload(error, overrides = nil, **kwargs)
      override_values = merge_hashes(overrides, kwargs)
      normalized = normalize_unknown_error(error, "Unknown error")
      merged_details = sanitize_json(
        merge_hashes(
          normalized[:details],
          get_alias(override_values, :details)
        )
      )
      merged_context = sanitize_json(build_context(override_values))

      payload = {
        "eventId" => coalesce(
          get_alias(override_values, :event_id, :eventId),
          SecureRandom.uuid
        ),
        "occurredAt" => coalesce(
          get_alias(override_values, :occurred_at, :occurredAt),
          Time.now.utc.iso8601
        ),
        "level" => coalesce(get_alias(override_values, :level), "error"),
        "platform" => coalesce(
          get_alias(override_values, :platform),
          @platform_name,
          "backend"
        ),
        "environment" => coalesce(get_alias(override_values, :environment), @environment),
        "service" => compact_object(
          sanitize_json(
            merge_hashes(
              {
                "name" => @service_name,
                "version" => @service_version,
                "instanceId" => @instance_id
              },
              get_alias(override_values, :service)
            )
          )
        ),
        "source" => compact_object(
          sanitize_json(
            merge_hashes(
              {
                "runtime" => detect_runtime,
                "language" => "ruby",
                "framework" => @framework,
                "component" => @component
              },
              get_alias(override_values, :source)
            )
          )
        ),
        "error" => compact_object(
          {
            "type" => coalesce(get_alias(override_values, :type), normalized[:error_type], "Exception"),
            "message" => coalesce(get_alias(override_values, :message), normalized[:message], "Unknown error"),
            "code" => coalesce(get_alias(override_values, :code), read_error_code(error)),
            "stack" => coalesce(get_alias(override_values, :stack), normalized[:stack]),
            "handled" => coalesce(get_alias(override_values, :handled), true),
            "details" => has_keys?(merged_details) ? merged_details : nil
          }
        ),
        "context" => has_keys?(merged_context) ? merged_context : nil
      }

      if @before_send
        next_payload = @before_send.call(payload)
        return nil if next_payload.nil? || next_payload == false

        payload = next_payload
      end

      prune_empty(compact_object(payload))
    end

    private

    def build_context(overrides)
      override_values = overrides || {}
      option_context = @context
      override_context = get_alias(override_values, :context) || {}
      tags = unique_strings(
        @tags +
        Array(get_alias(option_context, :tags)) +
        Array(get_alias(override_context, :tags)) +
        Array(get_alias(override_values, :tags))
      )

      prune_empty(
        merge_hashes(
          option_context,
          override_context,
          {
            request: merge_hashes(
              get_alias(option_context, :request),
              get_alias(override_context, :request),
              get_alias(override_values, :request)
            ),
            user: merge_hashes(
              get_alias(option_context, :user),
              get_alias(override_context, :user),
              get_alias(override_values, :user)
            ),
            session: merge_hashes(
              get_alias(option_context, :session),
              get_alias(override_context, :session),
              get_alias(override_values, :session)
            ),
            device: merge_hashes(
              get_alias(option_context, :device),
              get_alias(override_context, :device),
              get_alias(override_values, :device)
            ),
            correlationId: coalesce(
              get_alias(override_values, :correlation_id, :correlationId),
              get_alias(override_context, :correlation_id, :correlationId),
              get_alias(option_context, :correlation_id, :correlationId)
            ),
            tags: tags.empty? ? nil : tags
          }
        )
      )
    end

    def deliver(payload)
      @transport.call(
        {
          endpoint: @endpoint,
          api_token: @api_token,
          apiToken: @api_token,
          payload: payload,
          headers: @headers.dup,
          timeout: @timeout
        }
      )
    end

    def register_uncaught_hook
      return if @uncaught_hook_registered

      @uncaught_hook_registered = true
      at_exit do
        next unless @installed && @capture_uncaught

        error = $ERROR_INFO
        next if error.nil? || error.is_a?(SystemExit) || error.is_a?(SignalException)

        send_automatic_exception(
          error,
          handled: false,
          level: "critical",
          platform: @platform_name || "backend",
          source: { runtime: detect_runtime }
        )
      end
    end

    def send_automatic_exception(error, **kwargs)
      payload = build_payload(error, nil, **kwargs)
      return if payload.nil?

      deliver(payload)
    rescue StandardError => transport_error
      if @on_transport_error
        @on_transport_error.call(transport_error)
      else
        warn("AuraKeeper failed to send error log. #{transport_error}")
      end
    end

    def default_transport(config)
      uri = URI.parse(config[:endpoint])
      http = Net::HTTP.new(uri.host, uri.port)
      http.use_ssl = uri.scheme == "https"
      http.open_timeout = config[:timeout]
      http.read_timeout = config[:timeout]

      request = Net::HTTP::Post.new(
        uri.request_uri.empty? ? "/" : uri.request_uri,
        {
          "Content-Type" => "application/json",
          "X-API-Token" => config[:api_token]
        }.merge(config[:headers] || {})
      )
      request.body = JSON.generate(config[:payload])

      response = http.request(request)
      status = response.code.to_i
      raw_body = response.body.to_s

      raise RuntimeError, "AuraKeeper request failed with status #{status}: #{raw_body}" unless status.between?(200, 299)
      return { "status" => status } if raw_body.empty?

      content_type = response["content-type"].to_s
      return JSON.parse(raw_body) if content_type.include?("application/json")

      { "status" => status, "body" => raw_body }
    rescue SocketError, SystemCallError, Timeout::Error, JSON::ParserError => e
      raise RuntimeError, "AuraKeeper request failed: #{e.message}"
    end

    def normalize_unknown_error(value, fallback_message)
      if value.is_a?(Exception)
        return {
          error_type: value.class.name || "Exception",
          message: value.message.to_s.empty? ? fallback_message : value.message.to_s,
          stack: (value.backtrace || []).join("\n"),
          details: nil
        }
      end

      if value.is_a?(Hash) && !get_alias(value, :message).to_s.empty?
        return {
          error_type: get_alias(value, :name) || "Error",
          message: get_alias(value, :message),
          stack: get_alias(value, :stack),
          details: sanitize_json(value)
        }
      end

      if value.is_a?(String) && !value.empty?
        return {
          error_type: "Error",
          message: value,
          stack: nil,
          details: nil
        }
      end

      {
        error_type: "Error",
        message: fallback_message,
        stack: nil,
        details: value.nil? ? nil : { originalValue: sanitize_json(value) }
      }
    end

    def read_error_code(error)
      if error.respond_to?(:code)
        code = error.code
        return code.to_s unless code.nil? || code.to_s.empty?
      end

      if error.is_a?(Hash)
        code = get_alias(error, :code)
        return code.to_s unless code.nil? || code.to_s.empty?
      end

      nil
    end

    def detect_runtime
      (defined?(RUBY_ENGINE) ? RUBY_ENGINE : "ruby").to_s.downcase
    end

    def get_alias(hash, *keys)
      return nil unless hash.is_a?(Hash)

      keys.each do |key|
        return hash[key] if hash.key?(key)

        string_key = key.to_s
        return hash[string_key] if hash.key?(string_key)

        symbol_key = string_key.to_sym
        return hash[symbol_key] if hash.key?(symbol_key)
      end

      nil
    end

    def merge_hashes(*hashes)
      hashes.each_with_object({}) do |hash, merged|
        next unless hash.is_a?(Hash)

        hash.each do |key, value|
          merged[key] = value
        end
      end
    end

    def compact_object(hash)
      hash.each_with_object({}) do |(key, value), compacted|
        compacted[key] = value unless value.nil?
      end
    end

    def prune_empty(value)
      case value
      when Hash
        value.each_with_object({}) do |(key, child), pruned|
          next if child.nil?

          next_child = prune_empty(child)
          next if next_child.respond_to?(:empty?) && next_child.empty?

          pruned[key] = next_child
        end
      when Array
        value.filter_map do |child|
          next if child.nil?

          next_child = prune_empty(child)
          next if next_child.respond_to?(:empty?) && next_child.empty?

          next_child
        end
      else
        value
      end
    end

    def sanitize_json(value, depth = 0)
      return "[Truncated]" if depth >= MAX_SANITIZE_DEPTH

      case value
      when nil, String, Integer, Float, TrueClass, FalseClass
        value
      when Symbol
        value.to_s
      when Time, DateTime, Date
        value.iso8601
      when Array
        value.map { |item| sanitize_json(item, depth + 1) }
      when Hash
        value.each_with_object({}) do |(key, child), result|
          result[key.to_s] = sanitize_json(child, depth + 1)
        end
      when Exception
        {
          "class" => value.class.name,
          "message" => value.message,
          "backtrace" => sanitize_json(value.backtrace, depth + 1)
        }
      else
        value.respond_to?(:to_h) ? sanitize_json(value.to_h, depth + 1) : value.to_s
      end
    end

    def has_keys?(value)
      value.is_a?(Hash) && !value.empty?
    end

    def unique_strings(values)
      values.each_with_object([]) do |value, result|
        next if value.nil?

        string_value = value.to_s
        next if string_value.empty? || result.include?(string_value)

        result << string_value
      end
    end

    def coalesce(*values)
      values.find { |value| !value.nil? }
    end

    def blank?(value)
      value.nil? || value.to_s.empty?
    end
  end
end
